### PR TITLE
docs: 1/3 enable Sphinx nitpicky [fabric]

### PR DESCRIPTION
**Pull request by Borda**
## What does this PR do?

Reaction to #18068 

> [nitpicky](https://www.sphinx-doc.org/en/master/usage/configuration.html#confval-nitpicky)
If true, Sphinx will warn about all references where the target cannot be found. Default is False. You can activate this mode temporarily using the [-n](https://www.sphinx-doc.org/en/master/man/sphinx-build.html#cmdoption-sphinx-build-n) command-line switch.

ref: https://github.com/optuna/optuna/issues/4213

As part of the changes we have fixed several broken references and not rendered typings in docs

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

cc @borda @carmocca @justusschock @awaelchli

### CODE REVIEW

**Comment by Borda** [comment]
So not sure how to tell sphinx mapping also, this just built package... :/

**Comment by Borda** [comment]
missing refs to torch is strange... `WARNING: py:class reference target not found: torch.ScriptModule`

**Comment by Borda** [comment]
it is strange that nitpicky reject `Tensor` of it is imported as `from torch import Tensor`
also seems that all our references to protected API are not recognised... 🤔

**Comment by Borda** [comment]
not sure why some API have been missing types annotations, see for example: https://lightning.ai/docs/fabric/latest/api/generated/lightning.fabric.strategies.SingleDeviceStrategy.html

File: .github/workflows/docs-build.yml [hunk 1]
```diff
@@ -7,7 +7,7 @@ on:
     tags: ["*"]
   # use this event type to share secrets with forks.
   # it's important that the PR head SHA is checked out to run the changes
-  pull_request_target:
+  pull_request:  # fixme: pull_request_target
```
Comment (Borda) [inline .github/workflows/docs-build.yml:10]:
```suggestion
  pull_request_target:
```

File: src/lightning/fabric/plugins/collectives/collective.py [hunk 2]
```diff
@@ -111,12 +110,12 @@ def destroy_group(cls, group: CollectibleGroup) -> None:
     def _convert_to_native_op(cls, op: str) -> Any:
         ...
 
-    def setup(self, **kwargs: Any) -> Self:
```
Comment (awaelchli) [inline src/lightning/fabric/plugins/collectives/collective.py:114]: Self is correct here

File: src/lightning/fabric/fabric.py [hunk 3]
```diff
@@ -347,7 +347,7 @@ def setup_dataloaders(
     def _setup_dataloader(
         self, dataloader: DataLoader, use_distributed_sampler: bool = True, move_to_device: bool = True
     ) -> DataLoader:
-        """Set up a single dataloader for accelerated training.
+        r"""Set up a single dataloader for accelerated training.
```
Comment (awaelchli) [inline src/lightning/fabric/fabric.py:350]: why is this added everywhere? Seems highly unnecessary?

File: src/lightning/app/utilities/packaging/build_config.py [hunk 4]
```diff
@@ -193,7 +191,7 @@ def to_dict(self) -> Dict:
         return {"__build_config__": asdict(self)}
 
     @classmethod
-    def from_dict(cls, d: Dict) -> Self:
```
Comment (awaelchli) [inline src/lightning/app/utilities/packaging/build_config.py:196]: Self is correct here

File: docs/source-pytorch/advanced/model_parallel/fsdp.rst [hunk 5]
```diff
@@ -160,7 +160,7 @@ We can specify a list of layer classes in the **wrapping policy** to inform FSDP
         # 3. Pass it to the FSDPStrategy object
         strategy = FSDPStrategy(auto_wrap_policy=policy)
 
-    PyTorch provides several of these functional policies under :mod:`torch.distributed.fsdp.wrap`.
```
Comment (awaelchli) [inline docs/source-pytorch/advanced/model_parallel/fsdp.rst:163]: This was correct


See https://github.com/pytorch/pytorch/blob/main/torch/distributed/fsdp/wrap.py

File: docs/source-fabric/advanced/model_parallel/fsdp.rst [hunk 6]
```diff
@@ -142,7 +142,7 @@ We can specify a list of layer classes in the **wrapping policy** to inform FSDP
         # 3. Pass it to the FSDPStrategy object
         strategy = FSDPStrategy(auto_wrap_policy=policy)
 
-    PyTorch provides several of these functional policies under :mod:`torch.distributed.fsdp.wrap`.
```
Comment (awaelchli) [inline docs/source-fabric/advanced/model_parallel/fsdp.rst:145]: This was correct

See [pytorch/pytorch@main/torch/distributed/fsdp/wrap.py](https://github.com/pytorch/pytorch/blob/main/torch/distributed/fsdp/wrap.py)

File: src/lightning/fabric/wrappers.py [hunk 7]
```diff
@@ -48,6 +48,8 @@ def __init__(self, optimizer: Optimizer, strategy: Strategy, callbacks: Optional
             optimizer: The optimizer to wrap
             strategy: Reference to the strategy for handling the optimizer step
 
+        :meta public:
```
Comment (awaelchli) [inline src/lightning/fabric/wrappers.py:51]: What does this do? It seems unnecessary. Let's keep proteced APIs hidden, and docs are just there for the developers.

File: src/lightning/fabric/plugins/collectives/collective.py (hunk 2 above)
Comment (Borda) [reply src/lightning/fabric/plugins/collectives/collective.py:114]: OK and could it be also the class name?

File: src/lightning/fabric/fabric.py (hunk 3 above)
Comment (Borda) [reply src/lightning/fabric/fabric.py:350]: was not sure how the RST syntax works outside strings which are not marked as to be RST

File: src/lightning/app/utilities/packaging/build_config.py (hunk 4 above)
Comment (Borda) [reply src/lightning/app/utilities/packaging/build_config.py:196]: OK and could it be also the class name as string?

File: docs/source-pytorch/advanced/model_parallel/fsdp.rst (hunk 5 above)
Comment (Borda) [reply docs/source-pytorch/advanced/model_parallel/fsdp.rst:163]: maybe by source code but not according generated docs: https://pytorch.org/docs/stable/fsdp.html

File: docs/source-fabric/advanced/model_parallel/fsdp.rst (hunk 6 above)
Comment (Borda) [reply docs/source-fabric/advanced/model_parallel/fsdp.rst:145]: The same, PyTroch does not generate this as a module...

File: src/lightning/fabric/wrappers.py (hunk 7 above)
Comment (Borda) [reply src/lightning/fabric/wrappers.py:51]: yeah, remove my bad...
```suggestion
```

File: src/lightning/fabric/fabric.py [hunk 8]
```diff
@@ -471,8 +471,8 @@ def to_device(self, obj: Any) -> Any:
         ...
 
     def to_device(self, obj: Union[nn.Module, Tensor, Any]) -> Union[nn.Module, Tensor, Any]:
-        """Move a :class:`torch.nn.Module` or a collection of tensors to the current device, if it is not already on
-        that device.
+        r"""Move a :class:`torch.nn.Module` or a collection of tensors to the current device, if it is not already
```
Comment (Borda) [inline src/lightning/fabric/fabric.py:474]:
```suggestion
        """Move a :class:`torch.nn.Module` or a collection of tensors to the current device, if it is not already
```

File: docs/source-fabric/advanced/model_parallel/fsdp.rst (hunk 6 above)
Comment (awaelchli) [reply docs/source-fabric/advanced/model_parallel/fsdp.rst:145]: Ok, then maybe we remove `:mod:` and just reference "torch.distributed.fsdp"

File: docs/source-fabric/api/utilities.rst [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+lightning.fabric.utilities
+##########################
+
+.. autofunction:: lightning.fabric.utilities.seed.seed_everything
+
+.. autofunction:: lightning.fabric.utilities.seed.pl_worker_init_function
+
+.. autoclass:: lightning.fabric.utilities.types.Optimizable
+
+.. autoclass:: lightning.fabric.utilities.types.Steppable
```
Comment (awaelchli) [inline docs/source-fabric/api/utilities.rst:15]: I would prefer not to expose these three in the API docs here. They are not meant to be used by the user directly. cc @carmocca

File: docs/source-fabric/conf.py [hunk 10]
```diff
@@ -254,7 +254,23 @@
     "python": ("https://docs.python.org/3", None),
     "torch": ("https://pytorch.org/docs/stable/", None),
     "pytorch_lightning": ("https://lightning.ai/docs/pytorch/stable/", None),
+    "lightning": ("https://lightning.ai/docs/app/stable//", None),
```
Comment (awaelchli) [inline docs/source-fabric/conf.py:257]:
```suggestion
    "lightning": ("https://lightning.ai/docs/app/stable/", None),
```

Comment (awaelchli) [reply docs/source-fabric/conf.py:257]: Why is this pointing to docs/app? Seems counter-intuitive. Why is intersphinx needed between lightning and lightning

File: src/lightning/app/utilities/packaging/build_config.py (hunk 4 above)
Comment (awaelchli) [reply src/lightning/app/utilities/packaging/build_config.py:196]: nvm sorry, Self was not correct here

File: src/lightning/fabric/plugins/collectives/collective.py (hunk 2 above)
Comment (awaelchli) [reply src/lightning/fabric/plugins/collectives/collective.py:114]: Self means the same object gets returned as the the method was called on. But "Collective" would mean any object of that class could be returned. So here "Self" is more accurate and therefore correct. I don't think this should be changed and "sphinx" is wrong here.

Also, why would sphinx care about types? We have mypy for this.

Comment (Borda) [reply src/lightning/fabric/plugins/collectives/collective.py:114]: it tries to link types to external docs so you can cross-link to the API in external docs
so good to revert to `Self` and set it as ignore class :)

File: docs/source-fabric/advanced/model_parallel/fsdp.rst (hunk 6 above)
Comment (Borda) [reply docs/source-fabric/advanced/model_parallel/fsdp.rst:145]: the updated `:mod:'torch.distributed.fsdp'` is fine

File: docs/source-fabric/api/utilities.rst (hunk 9 above)
Comment (Borda) [reply docs/source-fabric/api/utilities.rst:15]: ok, then you can't refer them in docs, shall I remove the cross-link?

File: docs/source-fabric/conf.py (hunk 10 above)
Comment (Borda) [reply docs/source-fabric/conf.py:257]: Because you are referring to docs `LightningApp` and `LightningWork` so shall we remove the reference?

File: docs/source-fabric/api/utilities.rst (hunk 9 above)
Comment (awaelchli) [reply docs/source-fabric/api/utilities.rst:15]: Since pl_worker_init_function is referenced in seed_everything, maybe leave this one. But Optimizable and Steppable can be removed I think.

File: docs/source-fabric/conf.py (hunk 10 above)
Comment (awaelchli) [reply docs/source-fabric/conf.py]: Better to remove, otherwise we get in trouble if we want to reference other things like `lightning.pytorch` etc.

File: docs/source-fabric/api/utilities.rst (hunk 9 above)
Comment (Borda) [inline docs/source-fabric/api/utilities.rst:15]:
```suggestion
```

File: docs/source-fabric/conf.py [hunk 11]
```diff
@@ -252,9 +252,26 @@
...
+    "typing_extensions": ("https://typing-extensions.readthedocs.io/en/stable/", None),
     "torch": ("https://pytorch.org/docs/stable/", None),
     "pytorch_lightning": ("https://lightning.ai/docs/pytorch/stable/", None),
+    "tensorboardX": ("https://tensorboardx.readthedocs.io/en/stable/", None),
 }
+nitpicky = True
+
+nitpick_ignore = [
+    ("py:class", "typing.Self"),
+    # Nitpick does not see protected nor private API
```
Comment (Borda) [inline docs/source-fabric/conf.py:264]:
```suggestion
    # Nitpick does not see protected nor private API
    ("py:class", "lightning.fabric.utilities.types.Optimizable"),
    ("py:class", "lightning.fabric.utilities.types.Steppable"),
```

File: docs/source-fabric/conf.py [hunk 12]
```diff
@@ -252,9 +252,28 @@
...
+    ("py:class", "lightning.fabric.utilities.types.Optimizable"),
+    ("py:class", "lightning.fabric.utilities.types.Steppable"),
```
Comment (Borda) [inline docs/source-fabric/conf.py:266]:
```suggestion
    # these are not generated with docs API ref
    ("py:class", "lightning.fabric.utilities.types.Optimizable"),
    ("py:class", "lightning.fabric.utilities.types.Steppable"),
    # Nitpick does not see protected or private API
```

**Review by awaelchli — APPROVED** [review]

File: .github/workflows/docs-build.yml [hunk 13]
```diff
@@ -35,6 +35,7 @@ env:
 
 jobs:
   docs-checks:
+    if: github.event.pull_request.draft == false
```
Comment (carmocca) [inline .github/workflows/docs-build.yml:38]: I find building docs on drafts useful

File: docs/source-fabric/advanced/model_parallel/fsdp.rst (hunk 6 above)
Comment (carmocca) [reply docs/source-fabric/advanced/model_parallel/fsdp.rst:145]: But that doesn't point to the wrapping utilities specifically

File: docs/source-pytorch/advanced/model_parallel/fsdp.rst (hunk 5 above)
Comment (carmocca) [reply docs/source-pytorch/advanced/model_parallel/fsdp.rst:163]: The fsdp link doesn't include the apis for the wrapping policies. I suggest that this is skipped in nitpicky and added back once PyTorch fixes their docs to include that module

**Review by carmocca — APPROVED** [review]

File: .github/workflows/docs-build.yml (hunk 13 above)
Comment (Borda) [reply .github/workflows/docs-build.yml:38]: The HTML build is still there, this is just for doctests and linkcheck

File: docs/source-fabric/advanced/model_parallel/fsdp.rst (hunk 6 above)
Comment (Borda) [reply docs/source-fabric/advanced/model_parallel/fsdp.rst:145]: Yes, unfortunately wrapping is in PT docs or you found it somewhere? Otherwise we need to add plain link to PT codebase....

File: docs/source-pytorch/advanced/model_parallel/fsdp.rst (hunk 5 above)
Comment (Borda) [reply docs/source-pytorch/advanced/model_parallel/fsdp.rst:163]: Do you think it was bug in their docs and they are going to fix it?
